sycl/include/CL/sycl/half_type.hpp

```diff
@@ -71,6 +73,115 @@ class half {
 private:
   uint16_t Buf;
 };
+
+} // namespace host_half_impl
+
+namespace half_impl {
+class half;
+
+// On the host side we cannot use OpenCL cl_half# types as an underlying type
+// for vec because they are actually defined as an integer type under the hood.
+// As a result half values will be converted to the integer and passed as a
+// kernel argument which is expected to be floating point number.
+template <int NumElements> struct half_vec {
+  alignas(detail::vector_alignment<half, NumElements>::value)
+    std::array<half, NumElements> s;
+};
+
+#ifdef __SYCL_DEVICE_ONLY__
+  using StorageT = _Float16;
+  using BIsRepresentationT = _Float16;
+
+  using Vec2StorageT = StorageT __attribute__((ext_vector_type(2)));
+  using Vec3StorageT = StorageT __attribute__((ext_vector_type(3)));
+  using Vec4StorageT = StorageT __attribute__((ext_vector_type(4)));
+  using Vec8StorageT = StorageT __attribute__((ext_vector_type(8)));
+  using Vec16StorageT = StorageT __attribute__((ext_vector_type(16)));
+#else
+  using StorageT = detail::host_half_impl::half;
+  // On host, built-in functions must operate on this generic implementation
```

Comment:
Actually, I don't remember exact reason why we can't use the same `StorageT` and `BIsRepresentationT` on host and I'm trying to unify them: so far I've faced a lot of issues with math built-ins implementation for host device, trying to resolve them.

Reply:
Ok, I spent some time investigating this and realized that there is no much sense in doing this, comparing to amount of changes required to implement this.

On device side, we have to map `half` to native `_Float16` to get it working with built-in functions. On host side, we anyway implemented these built-ins and `half` is not represented by some native type, but also by some custom implementation via C++ class. Having that said, there are no benefits in unwrapping host half implementation.

I tried to explain this in comments in the code